OpenRA.Mods.D2k/Traits/Sandworm.cs

```diff
@@ -0,0 +1,132 @@
+#region Copyright & License Information
+/*
+* Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
+* This file is part of OpenRA, which is free software. It is made
+* available to you under the terms of the GNU General Public License
+* as published by the Free Software Foundation. For more information,
+* see COPYING.
+*/
+#endregion
+
+using System;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.D2k.Traits
+{
+	class SandwormInfo : WandersInfo
+	{
+		public const string IdleSequenceName = "idle";
+
+		[Desc("Time between rescanning for targets (in ticks).")]
+		public const int TargetRescanIntervalInTicks = 32;
+
+		[Desc("The radius in which the worm \"searches\" for targets.")]
+		public readonly int MaxSearchRadius = 32;
+
+		[Desc("The range at which the worm launches an attack regardless of noise levels.")]
+		public readonly int LauchAttackRange = 3;
+
+		public override object Create(ActorInitializer init) { return new Sandworm(init.Self, this); }
+	}
+
+	class Sandworm : Wanders, ITick, INotifyBecomingIdle
+	{
+		readonly Lazy<IMove> move;
+		readonly Lazy<RenderUnit> renderUnit;
+		readonly Lazy<AttackBase> attackTrait;
+		readonly Lazy<IPositionable> positionable; 
+		readonly SandwormInfo info;
+
+		public bool IsAttacking;
+
+		int targetCountdown;
+
+		public Sandworm(Actor self, SandwormInfo info)
+			: base(self, info)
+		{
+			this.info = info;
+			move = Exts.Lazy(self.Trait<IMove>);
+			renderUnit = Exts.Lazy(self.Trait<RenderUnit>);
+			attackTrait = Exts.Lazy(self.Trait<AttackBase>);
+			positionable = Exts.Lazy(self.Trait<IPositionable>);
+		}
+
+		public new void OnBecomingIdle(Actor self)
+		{
+			if (renderUnit.Value.DefaultAnimation.CurrentSequence.Name != SandwormInfo.IdleSequenceName)
+				renderUnit.Value.DefaultAnimation.ReplaceAnim("idle");
+
+			base.OnBecomingIdle(self);
+		}
+
+		public override void DoAction(Actor self, CPos targetCell)
+		{
+			if (CheckForTargets(self))
+				return;
+
+			Game.Debug("(Sandworm {0}) is wandering.", self.ActorID);
+			self.QueueActivity(move.Value.MoveWithinRange(Target.FromCell(self.World, targetCell, SubCell.Any), WRange.FromCells(1)));
+		}
+
+		public void Tick(Actor self)
+		{
+			if (--targetCountdown > 0 || IsAttacking || !self.IsInWorld)
+				return;
+
+			CheckForTargets(self);
+		}
+
+		private bool CheckForTargets(Actor self)
```

Comment:
This needs a better name.
`CheckForTargets` is a verb, something like `AnyValidTargets` would be better.